encoding/ccf/decode.go

```diff
@@ -0,0 +1,1887 @@
+	// Decode array length.
+	n, err := d.dec.DecodeArrayHead()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if number of elements is even.
+	if n%2 != 0 {
+		return nil, fmt.Errorf(
+			"encoded dict-value has %d elements (expected even number of elements)",
+			n,
+		)
+	}
+
+	// previousKeyRawBytes is used to determine if dictionary keys are sorted
+	var previousKeyRawBytes []byte
+
+	pairCount := n / 2
+	pairs := make([]cadence.KeyValuePair, pairCount)
+	for i := 0; i < int(pairCount); i++ {
+		// element i: key
+
+		// Decode key as raw bytes to check that key pairs are sorted by key.
+		keyRawBytes, err := d.dec.DecodeRawBytes()
+		if err != nil {
+			return nil, err
+		}
+
+		// "Deterministic CCF Encoding Requirements" in CCF specs:
+		//
+		//   "dict-value key-value pairs MUST be sorted by key."
+		if !bytesAreSortedBytewise(previousKeyRawBytes, keyRawBytes) {
+			return nil, fmt.Errorf("encoded dict-value keys are not sorted")
+		}
+
+		previousKeyRawBytes = keyRawBytes
+
+		// decode key from raw bytes
+		keyDecoder := NewDecoder(d.gauge, keyRawBytes)
+		key, err := keyDecoder.decodeValue(typ.KeyType, types)
+		if err != nil {
+			return nil, err
+		}
+
+		// element i+1: value
+		element, err := d.decodeValue(typ.ElementType, types)
+		if err != nil {
+			return nil, err
+		}
+
+		pairs[i] = cadence.NewMeteredKeyValuePair(d.gauge, key, element)
+	}
+
+	value, err := cadence.NewMeteredDictionary(
+		d.gauge,
+		len(pairs),
+		func() ([]cadence.KeyValuePair, error) {
+			// "Valid CCF Encoding Requirements" in CCF specs:
+			//
+			//   "Keys MUST be unique in dict-value. Decoders are not always required to check
+			//   for duplicate dictionary keys. In some cases, checking for duplicate dictionary
+			//   key is not necessary or it may be delegated to the application."
+			//
+			// Here, decoder doesn't check uniqueness of dictionary keys
+			// because checking is delegated (entrusted) to Cadence runtime.
+			return pairs, nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return value.WithType(typ), nil
+}
+
+// decodeComposite decodes encoded composite-value as
+// language=CDDL
+// composite-value = [+ (field: value)]
+func (d *Decoder) decodeComposite(fieldTypes []cadence.Field, types cadenceTypeByCCFTypeID) ([]cadence.Value, error) {
+	fieldCount := len(fieldTypes)
+
+	// Decode number of fields.
+	err := decodeCBORArrayWithKnownSize(d.dec, uint64(fieldCount))
+	if err != nil {
+		return nil, err
+	}
+
+	common.UseMemory(d.gauge, common.MemoryUsage{
+		Kind:   common.MemoryKindCadenceField,
+		Amount: uint64(fieldCount),
+	})
+
+	fieldValues := make([]cadence.Value, fieldCount)
+
+	for i := 0; i < fieldCount; i++ {
+		// Decode field.
+		field, err := d.decodeValue(fieldTypes[i].Type, types)
+		if err != nil {
+			return nil, err
+		}
+		fieldValues[i] = field
+	}
+
+	return fieldValues, nil
+}
+
+// decodeStruct decodes encoded composite-value as
+// language=CDDL
+// composite-value = [+ (field: value)]
```

Comment:
Here and for the other composite value decoding functions: Should this be "zero or more"? Maybe also check encoding function. Might also be an issue in the spec 

```suggestion
// composite-value = [* (field: value)]
```

Reply:
Done in commits f89eb0615b95c5bb5deeeed7966fa0d6fa98e9d0 and 42d865792c152b0af33991e4ad5d1b6adf796249.  

Codec now matches today's updates to CCF specs.  Changes to CCF specs include:
- https://github.com/fxamacker/ccf_draft/pull/66
- https://github.com/fxamacker/ccf_draft/pull/67